zipkin/src/main/java/zipkin/storage/InMemorySpanStore.java

```diff
@@ -40,12 +39,34 @@
 
 /** Internally, spans are indexed on 64-bit trace ID */
 public final class InMemorySpanStore implements SpanStore {
-  private final Multimap<Long, Span> traceIdToSpans = new LinkedListMultimap<>();
-  private final Set<Pair<Long>> traceIdTimeStamps = new TreeSet<>(VALUE_2_DESCENDING);
-  private final Multimap<String, Pair<Long>> serviceToTraceIdTimeStamp =
-      new SortedByValue2Descending<>();
-  private final Multimap<String, String> serviceToSpanNames =
-      new LinkedHashSetMultimap<>();
+  /**
+   * Primary source of data is this map, which includes spans ordered descending by timestamp. All
+   * other maps are derived from the span values here. This uses a list for the spans, so that it is
+   * visible (via /api/v1/trace/id?raw) when instrumentation report the same spans multiple times.
```

Comment:
@joel-airspring added this note, as clearly I forgot why when you asked earlier. ZipkinRuleTest reminded me :)